aws-api/src/main/java/com/amplifyframework/api/aws/GsonErrorDeserializer.java

```diff
@@ -0,0 +1,112 @@
+
+import com.amplifyframework.api.graphql.GraphQLResponse;
+import com.amplifyframework.api.graphql.error.GraphQLLocation;
+import com.amplifyframework.api.graphql.error.GraphQLPathSegment;
+
+import com.google.gson.JsonDeserializationContext;
+import com.google.gson.JsonDeserializer;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParseException;
+import com.google.gson.JsonPrimitive;
+import com.google.gson.reflect.TypeToken;
+
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+public final class GsonErrorDeserializer implements JsonDeserializer<GraphQLResponse.Error> {
+    private static final String MESSAGE_KEY = "message";
+    private static final String LOCATIONS_KEY = "locations";
+    private static final String PATH_KEY = "path";
+    private static final String EXTENSIONS_KEY = "extensions";
+
+    @Override
+    public GraphQLResponse.Error deserialize(JsonElement json,
+                 Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
+        if (!(json.isJsonObject())) {
+            throw new JsonParseException("Expected a JSONObject but found a " +
+                    json.getClass().getName() + " while deserializing error");
+        }
+
+        String message = null;
+        List<GraphQLLocation> locations = null;
+        List<GraphQLPathSegment> path = null;
+        Map<String, Object> extensions = null;
+        JsonObject extensionsJson = new JsonObject();
+        JsonObject nonSpecifiedData = new JsonObject();
+
+        JsonObject error = json.getAsJsonObject();
+
+        for (String key : error.keySet()) {
+            JsonElement value = error.get(key);
+            if (value != null && !value.isJsonNull()) {
+                if (key.equals(MESSAGE_KEY)) {
+                    message = context.deserialize(value, String.class);
+                } else if (key.equals(LOCATIONS_KEY)) {
+                    Type locationsType = new TypeToken<List<GraphQLLocation>>() { }.getType();
+                    locations = context.deserialize(value, locationsType);
+                } else if (key.equals(PATH_KEY)) {
+                    path = getPath(value);
+                } else if (key.equals(EXTENSIONS_KEY)) {
+                    extensionsJson = value.getAsJsonObject();
+                } else {
+                    nonSpecifiedData.add(key, value);
+                }
+            }
+        }
+
+        // Merge nonSpecifiedData into extensions, but don't overwrite anything already there.
+        for (String key : nonSpecifiedData.keySet()) {
+            if (!extensionsJson.has(key)) {
+                extensionsJson.add(key, nonSpecifiedData.get(key));
+            }
+        }
+
+        // Deserialize the extensions JSON to a Map
+        if (extensionsJson.size() > 0) {
+            Type extensionType = new TypeToken<Map<String, Object>>() {
```

Comment:
You can add a checkstyle suppression if you want to keep this `{}` together. Or, we could update the checkstyle rules to not enforce this rule for class bodies.

I think this is [`WhitespaceAround`](https://github.com/aws-amplify/amplify-android/blob/master/configuration/checkstyle-rules.xml#L91) right? Might be addressed by adding a property for [`allowEmptyTypes`](https://checkstyle.sourceforge.io/config_whitespace.html#WhitespaceAround_Properties) as `true`?

Reply:
Thanks for the tip!  I updated `checkstyle-rules.xml`